File: C/210-course-schedule-ii/main.go
package main

import (
	"fmt"
)

func findOrder(numCourses int, prerequisites [][]int) []int {
	M := make(map[int][]int)
	E := make(map[int]int)
	for i := 0; i < numCourses; i++ {
		E[i] = 0
	}
	for i := 0; i < len(prerequisites); i++ {
		p := prerequisites[i]
		v, ok := M[p[1]]
		if !ok {
			M[p[1]] = []int{p[0]}
		} else {
			M[p[1]] = append(v, p[0])
		}
		E[p[0]]++
	}

	findZeroIngress := func(E map[int]int) []int {
		var ret []int
		for k, v := range E {
			if v == 0 {
				ret = append(ret, k)
			}
		}
		return ret
	}

	var ret [][]int
	idxArr := findZeroIngress(E)
	for len(idxArr) > 0 {
		ret = append(ret, idxArr)
		for i := 0; i < len(idxArr); i++ {
			numCourses--
			if v, ok := M[idxArr[i]]; ok {
				for j := 0; j < len(v); j++ {
					E[v[j]]--
				}
			}
			delete(M, idxArr[i])
			delete(E, idxArr[i])
		}
		idxArr = findZeroIngress(E)
	}
	if numCourses > 0 {
		return nil
	}

	var r []int
	for i := 0; i < len(ret); i++ {
		r = append(r, ret[i]...)
	}
	return r
}

func main() {
	fmt.Println(findOrder(2, [][]int{[]int{0, 1}}))
}

File: C/210-course-schedule-ii/main.rs
struct Solution {}

use std::collections::HashMap;
impl Solution {
  pub fn find_order(num_courses: i32, prerequisites: Vec<Vec<i32>>) -> Vec<i32> {
    let mut ingress: HashMap<i32, Vec<i32>> = HashMap::new();
    let mut out: HashMap<i32, Vec<i32>> = HashMap::new();

    (0..num_courses).for_each(|i| {
      ingress.insert(i, Vec::new());
      out.insert(i, Vec::new());
    });

    prerequisites.iter().for_each(|p| {
      ingress.get_mut(&p[0]).unwrap().push(p[1]);
      out.get_mut(&p[1]).unwrap().push(p[0]);
    });

    let mut result: Vec<i32> = Vec::new();

    loop {
      let mut found = false;
      let mut candi: i32 = -1;

      for (k, v) in ingress.iter() {
        if v.len() == 0 {
          found = true;
          result.push(*k);
          candi = *k;
          break;
        }
      }
      if !found {
        break;
      }
      ingress.remove(&candi);
      for i in out.get(&candi).unwrap() {
        ingress.get_mut(i).unwrap().retain(|x| *x != candi);
      }
    }

    if ingress.len() > 0 {
      return Vec::new();
    }
    result
  }
}
